tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLog.java

```diff
@@ -107,45 +119,100 @@ public TransactionLogReader getReader() throws IOException {
+    private List<Entry> transactionEntries;
     public LogWriter(FileSystem fs, Configuration hConf, Path logPath) throws IOException {
       // TODO: retry a few times to ride over transient failures?
+      SequenceFile.Metadata metadata = new SequenceFile.Metadata();
+      metadata.set(new Text("version"), new Text("v2"));
+
       this.internalWriter =
-        SequenceFile.createWriter(fs, hConf, logPath, LongWritable.class, TransactionEdit.class);
+        SequenceFile.createWriter(fs, hConf, logPath, LongWritable.class,
+                                  TransactionEdit.class, SequenceFile.CompressionType.NONE, null, null, metadata);
+      numEntries = 0;
+      transactionEntries = new ArrayList<>();
+
       LOG.debug("Created a new TransactionLog writer for " + logPath);
     }
 
     @Override
     public void append(Entry entry) throws IOException {
-      internalWriter.append(entry.getKey(), entry.getEdit());
+      transactionEntries.add(entry);
     }
 
     @Override
     public void sync() throws IOException {
+      // write the number of entries we are writing to the log
+      if (transactionEntries.size() > 0) {
+        String key = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED;
+        internalWriter.appendRaw(key.getBytes(), 0, key.getBytes().length, new SequenceFile.ValueBytes() {
+          @Override
+          public void writeUncompressedBytes(DataOutputStream outStream) throws IOException {
+            outStream.write(toBytes(transactionEntries.size()));
+            outStream.flush();
+          }
+
+          @Override
+          public void writeCompressedBytes(DataOutputStream outStream) throws IllegalArgumentException, IOException {
+            throw new IllegalArgumentException("UncompressedBytes cannot be compressed!");
+          }
+
+          @Override
+          public int getSize() {
+            // size of value, which is an integer
+            return Integer.SIZE / Byte.SIZE;
+          }
+        });
+      }
+
+      // write the entries to the log
+      for (Entry entry : transactionEntries) {
+        internalWriter.append(entry.getKey(), entry.getEdit());
+      }
+      transactionEntries.clear();
       internalWriter.syncFs();
     }
 
     @Override
     public void close() throws IOException {
+      transactionEntries.clear();
       internalWriter.close();
     }
   }
 
   private static final class LogReader implements TransactionLogReader {
-
     private boolean closed;
     private SequenceFile.Reader reader;
     private LongWritable key = new LongWritable();
+    private boolean newVersion;
+    private List<TransactionEdit> transactionEdits;
 
     public LogReader(SequenceFile.Reader reader) {
       this.reader = reader;
+      newVersion = reader.getMetadata().getMetadata().containsKey(new Text("version"));
+      transactionEdits = new ArrayList<>();
     }
 
     @Override
     public TransactionEdit next() {
       try {
+        if (newVersion) {
```

Comment:
Instead of using checks for newVersion in Reader everywhere, can we have ReaderV2 and ReaderV1. Based on the version in metadata we can select the right reader to use. This will keep code simple. See if we can re-use code between the readers. Take a look at TransactionEditCodec hierarchy.